From 4470b3876478ffa31355948c01dcd2cafe272eed Mon Sep 17 00:00:00 2001
From: Hongzhang Yang <Hongzhang.Yang@freescale.com>
Date: Thu, 25 Oct 2012 15:21:25 +0800
Subject: [PATCH 219/280] ENGR00231107 VPU lib workaround for VPU BWB blocking
 issue

VPU lib 5.4.7
- Bug description:

When playing some VC1 clip with Android Gallery, sometimes VPU is
blocked in waiting for BWB done. PC = {0407,0409}

Rigel, FW v2.1.7_r33576

Clip can be found at:
\\10.192.225.215\01_CodecVectors\SHAVectors\VC1Dec\Conformance\1080\WMV9_MPHL_1920x1080_23.98fps_7753kbps_WMA9.1pro_6ch_MiamiVice.wmv

It's linked to original ticket ENGR00223231
FSLARIK-278 [Decoder/VC1] VPU was blocked in waiting for BWB done

- Workaround

Disable BWB for VC1 decoder
The bug can only be reproduced on Android. CnM is checking it.

Signed-off-by: Hongzhang Yang <Hongzhang.Yang@freescale.com>
---
 vpu/vpu_lib.c | 4 ++++
 vpu/vpu_lib.h | 2 +-
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/vpu/vpu_lib.c b/vpu/vpu_lib.c
index ad72495..c4b9ecb 100644
--- a/vpu/vpu_lib.c
+++ b/vpu/vpu_lib.c
@@ -2616,6 +2616,10 @@ RetCode vpu_DecOpen(DecHandle * pHandle, DecOpenParam * pop)
 		if (pDecInfo->mapType)
 			val |= (pDecInfo->tiledLinearEnable << 11 | 0x03 << 9);
 		val |= 1 << 12;
+
+		// workaround for BWB blocking issue (ENGR00231107)
+		if (pDecInfo->openParam.bitstreamFormat == STD_VC1)
+			val &= ~(1 << 12);
 	}
 	pCodecInst->ctxRegs[CTX_BIT_FRAME_MEM_CTRL] =
 		    val | (pDecInfo->openParam.chromaInterleave << 2);
diff --git a/vpu/vpu_lib.h b/vpu/vpu_lib.h
index 6c6b898..a992af8 100644
--- a/vpu/vpu_lib.h
+++ b/vpu/vpu_lib.h
@@ -667,7 +667,7 @@ typedef struct vpu_versioninfo {
  * v4.2.2 [2008.09.03] support encoder on MX51
  * v4.0.2 [2008.08.21] add the IOClkGateSet() for power saving.
  */
-#define VPU_LIB_VERSION_CODE	VPU_LIB_VERSION(5, 4, 6)
+#define VPU_LIB_VERSION_CODE	VPU_LIB_VERSION(5, 4, 7)
 
 extern unsigned int system_rev;
 
-- 
1.8.0

